### PR TITLE
Generalize homogenized thermal conductivity

### LINKED ISSUE #22919: Generalize computation of homogenized thermal conductivities
## Reason

Enable computation of off-diagonal elements in homogenized thermal conductivity tensor
and allow using anisotropic thermal conductivity as input.

## Design

Extension of `HomogenizedThermalConductivity`.

## Impact

Additional capability, some changes in input and documentation.

### PR DESCRIPTION
**Pull request by snschune**
## Reason

To compute homogenized thermal conductivities in non-square lattices,
we need to compute off-diagonal thermal conductivity components.
The current homogenization object do not support that.

In addition, the documentation needed some clarification and fixing
of a sign error. 

Finally, I want to make the homogenization work with anisotropic input
thermal conductivities.

## Design

Refactor `HomogenizedThermalConductivity` and clean up its input parameters.
Add anisotropic object for  `HomogenizedHeatConduction` (not complete yet).

## Impact

Some changes to input parameters. I think it's actually clearer now.

closes #22919

### CODE REVIEW

**Comment by snschune** [comment]
@grmnptr this is where I would like your help to check the math. It should be in the documentation + the code + the reference I added.

**Review by grmnptr — CHANGES_REQUESTED** [review]
Great work! Just a few questions/ideas.

File: modules/heat_conduction/doc/content/source/kernels/AnisoHomogenizedHeatConduction.md [hunk 1]
```diff
@@ -0,0 +1,26 @@
+# AnisoHomogenizedHeatConduction
+
+!syntax description /Kernels/AnisoHomogenizedHeatConduction
+
+## Description
+
+This `Kernel` computes the right hand side of the equation
+
+!equation
+\int_Y \frac{\partial v}{\partial y_i}k_{ij} \frac{\partial \chi^k}{\partial y_j} = -\int_Y \frac{\partial v}{\partial y_i}k_{ik} \text{d}y
```
[the remark this answers is not in the document]
Comment (snschune) [reply modules/heat_conduction/doc/content/source/kernels/AnisoHomogenizedHeatConduction.md:10]: `k` is already in use for thermal conductivity. Should use `\lambda` for thermal conductivity instead.

File: modules/heat_conduction/doc/content/source/postprocessors/HomogenizedThermalConductivity.md [hunk 2]
```diff
@@ -5,14 +5,25 @@
...
+!equation id=eq:1
+\int_Y \frac{\partial v}{\partial y_i}k_{ij} \frac{\partial \chi^k}{\partial y_j} = -\int_Y \frac{\partial v}{\partial y_i}k_{ik} \text{d}y
+
+
+This `PostProcessor` is used in conjunction with the [Heat Conduction](HeatConduction.md) `Kernel` and the [Homogenized Heat Conduction](HomogenizedHeatConduction.md) `Kernel`.
+An application can be found in [!cite](hales15homogenization).
+
+!alert note title=Notation
+First, compared to standard notation for homogenization theory applied to thermal conductivity (e.g. [!cite](SONG2006710)),
+MOOSE computes $-\chi^j$. This is achieved by inverting the sign of the right hand side of [eq:1]. This leads to an inversion of the sign in the parenthesis in [eq:0].
```
[the remark this answers is not in the document]
Comment (snschune) [reply modules/heat_conduction/doc/content/source/postprocessors/HomogenizedThermalConductivity.md:23]: Could do that, but it is handled internally so it's not a big danger that users get that wrong.

File: modules/heat_conduction/src/kernels/AnisoHomogenizedHeatConduction.C [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+Real
+AnisoHomogenizedHeatConduction::computeQpResidual()
+{
+  // This is the matrix-vector product of the diffusion coefficient tensor with
+  // the j-th unit vector
+  RealVectorValue d_times_ej(0, 0, 0);
+  for (unsigned int j = 0; j < _mesh.dimension(); ++j)
+    d_times_ej(j) = _diffusion_coefficient[_qp](j, _component);
+  return d_times_ej * _grad_test[_i][_qp];
```
[the remark this answers is not in the document]
Comment (snschune) [reply modules/heat_conduction/src/kernels/AnisoHomogenizedHeatConduction.C:48]: In the non-INL papers, the rhs is always negative. However, we are solving for `-chi`, so our rhs is positive. Then we multiply by a test function and "integrate by parts" which switches the sign (so it's negative then), and then we bring it over to the left hand side. So you have two sign changes and the above is correct.

File: modules/heat_conduction/src/postprocessors/HomogenizedThermalConductivity.C [hunk 4]
```diff
@@ -16,33 +16,60 @@ registerMooseObject("HeatConductionApp", HomogenizedThermalConductivity);
...
+    _tensor_diffusion_coefficient = &getMaterialProperty<RankTwoTensor>("diffusion_coefficient");
+    _diffusion_coefficient = nullptr;
+  }
+  else
+  {
+    _diffusion_coefficient = &getMaterialProperty<Real>("diffusion_coefficient");
+    _tensor_diffusion_coefficient = nullptr;
+  }
+
+  if (_row >= _dim)
```
[the remark this answers is not in the document]
Comment (snschune) [reply modules/heat_conduction/src/postprocessors/HomogenizedThermalConductivity.C:61]: You can have off-diagonal diffusion coefficients if the heterogeneous coefficient is a scalar. That is for example the case in hexagonal geometries...i.e. where the principal axes are not e_x and e_y.

**Review by grmnptr — APPROVED** [review]
Great job!

**Pull request merged**